File: obc_controller/ui/graph_panel.py
"""Real-time graph panel using pyqtgraph — dark space theme.

Three separate sub-plots with linked X axes:
  1. Voltage: Vout and Vin (V)
  2. Current: Iout (A)
  3. Temperature: Temp (deg C)

Supports event markers (vertical lines + labels) across all plots.
"""

from __future__ import annotations

import csv
import time
from collections import deque

import numpy as np
import pyqtgraph as pg
from PySide6.QtCore import Qt, QTimer
from PySide6.QtGui import QFont
from PySide6.QtWidgets import (
    QComboBox,
    QFileDialog,
    QGroupBox,
    QHBoxLayout,
    QLabel,
    QPushButton,
    QVBoxLayout,
)

from obc_controller.can_protocol import Message2
from obc_controller.ui.theme import (
    BG_DEEP,
    BORDER,
    CYAN,
    GREEN,
    MAGENTA,
    ORANGE,
    RED,
    TEXT_DIM,
)

# Time-window options (minutes)
WINDOW_OPTIONS = {
    "1 min": 60,
    "5 min": 300,
    "10 min": 600,
    "30 min": 1800,
}

# Severity → line colour
_SEV_COLORS = {
    "info": CYAN,
    "warning": ORANGE,
    "error": RED,
}

# Apply pyqtgraph global dark config
pg.setConfigOptions(
    antialias=True,
    background=BG_DEEP,
    foreground=TEXT_DIM,
)

_MARKER_FONT = QFont("sans-serif", 8)


class GraphPanel(QGroupBox):
    def __init__(self, parent=None):
        super().__init__("Live Graph", parent)

        self._t0 = time.monotonic()
        self._paused = False

        # Ring buffers (store up to 30 min @ 2 Hz = 3600 points max)
        maxlen = 3600
        self._ts: deque[float] = deque(maxlen=maxlen)
        self._vout: deque[float] = deque(maxlen=maxlen)
        self._vin: deque[float] = deque(maxlen=maxlen)
        self._iout: deque[float] = deque(maxlen=maxlen)
        self._temp: deque[float] = deque(maxlen=maxlen)
        self._status: deque[int] = deque(maxlen=maxlen)

        self._window_sec = 600  # default 10 min
        self._data_dirty = False  # flag to avoid redundant np.array() calls

        # Event markers: [(timestamp, label, severity, [lines], text_item)]
        self._markers: list[tuple] = []
        self._marker_data: list[tuple[float, str, str]] = []

        layout = QVBoxLayout(self)

        # --- controls row ---
        ctrl_row = QHBoxLayout()
        self._pause_btn = QPushButton("Pause")
        self._pause_btn.setCheckable(True)
        self._pause_btn.toggled.connect(self._on_pause_toggled)
        ctrl_row.addWidget(self._pause_btn)

        self._clear_btn = QPushButton("Clear")
        self._clear_btn.clicked.connect(self._clear_data)
        ctrl_row.addWidget(self._clear_btn)

        ctrl_row.addWidget(QLabel("Window:"))
        self._window_combo = QComboBox()
        self._window_combo.addItems(list(WINDOW_OPTIONS.keys()))
        self._window_combo.setCurrentText("10 min")
        self._window_combo.currentTextChanged.connect(self._on_window_changed)
        ctrl_row.addWidget(self._window_combo)

        self._export_btn = QPushButton("Export CSV")
        self._export_btn.clicked.connect(self._export_csv)
        ctrl_row.addWidget(self._export_btn)

        ctrl_row.addStretch()
        layout.addLayout(ctrl_row)

        # --- pyqtgraph widget with 3 rows ---
        self._graphics = pg.GraphicsLayoutWidget()
        layout.addWidget(self._graphics)

        grid_pen = pg.mkPen(BORDER, width=0.5)

        # Plot 1: Voltage (Vout + Vin)
        self._p_volt = self._graphics.addPlot(row=0, col=0, title="Voltage")
        self._p_volt.setLabel("left", "V", color=TEXT_DIM)
        self._p_volt.setLabel("bottom", "s", color=TEXT_DIM)
        self._p_volt.addLegend(brush=(17, 24, 39, 180))
        self._p_volt.showGrid(x=True, y=True, alpha=0.15)
        self._p_volt.getAxis("bottom").setPen(grid_pen)
        self._p_volt.getAxis("left").setPen(grid_pen)
        self._curve_vout = self._p_volt.plot(
            pen=pg.mkPen(CYAN, width=2), name="Vout"
        )
        self._curve_vin = self._p_volt.plot(
            pen=pg.mkPen(MAGENTA, width=2), name="Vin"
        )

        # Plot 2: Current (Iout only)
        self._p_curr = self._graphics.addPlot(row=1, col=0, title="Current")
        self._p_curr.setLabel("left", "A", color=TEXT_DIM)
        self._p_curr.setLabel("bottom", "s", color=TEXT_DIM)
        self._p_curr.addLegend(brush=(17, 24, 39, 180))
        self._p_curr.showGrid(x=True, y=True, alpha=0.15)
        self._p_curr.getAxis("bottom").setPen(grid_pen)
        self._p_curr.getAxis("left").setPen(grid_pen)
        self._curve_iout = self._p_curr.plot(
            pen=pg.mkPen(GREEN, width=2), name="Iout"
        )

        # Plot 3: Temperature
        self._p_temp = self._graphics.addPlot(
            row=2, col=0, title="Temperature"
        )
        self._p_temp.setLabel("left", "\u00b0C", color=TEXT_DIM)
        self._p_temp.setLabel("bottom", "s", color=TEXT_DIM)
        self._p_temp.addLegend(brush=(17, 24, 39, 180))
        self._p_temp.showGrid(x=True, y=True, alpha=0.15)
        self._p_temp.getAxis("bottom").setPen(grid_pen)
        self._p_temp.getAxis("left").setPen(grid_pen)
        self._curve_temp = self._p_temp.plot(
            pen=pg.mkPen(ORANGE, width=2), name="Temp"
        )

        # Link X axes
        self._p_curr.setXLink(self._p_volt)
        self._p_temp.setXLink(self._p_volt)

        # All three plots for marker iteration
        self._plots = (self._p_volt, self._p_curr, self._p_temp)

        # Redraw timer (~15 Hz)
        self._redraw_timer = QTimer(self)
        self._redraw_timer.timeout.connect(self._redraw)
        self._redraw_timer.start(66)

    # ---- public API -------------------------------------------------------

    def add_point(self, msg: Message2) -> None:
        t = time.monotonic() - self._t0
        self._ts.append(t)
        self._vout.append(msg.output_voltage)
        self._vin.append(msg.input_voltage)
        self._iout.append(msg.output_current)
        self._temp.append(msg.temperature)
        self._status.append(msg.status.to_byte())
        self._data_dirty = True

    def add_event_marker(
        self, label: str, severity: str = "info"
    ) -> None:
        """Add a vertical event marker across all 3 plots.

        *severity* is one of ``"info"``, ``"warning"``, ``"error"``.
        The marker line + label is drawn on all three plots (X-linked).
        """
        t = time.monotonic() - self._t0
        color = _SEV_COLORS.get(severity, CYAN)
        pen = pg.mkPen(color, width=1.5, style=Qt.PenStyle.DashLine)

        lines: list = []
        for plot in self._plots:
            line = pg.InfiniteLine(pos=t, angle=90, pen=pen)
            plot.addItem(line)
            lines.append(line)

        # Text label on the voltage plot (top)
        text_item = pg.TextItem(text=label, color=color, anchor=(0, 0))
        text_item.setFont(_MARKER_FONT)
        text_item.setPos(t, 0)
        self._p_volt.addItem(text_item)

        self._markers.append((t, label, severity, lines, text_item))
        self._marker_data.append((t, label, severity))

    # ---- internal ---------------------------------------------------------

    def _redraw(self) -> None:
        if self._paused or len(self._ts) == 0:
            return

        if not self._data_dirty:
            return
        self._data_dirty = False

        ts = np.array(self._ts)
        now = ts[-1]
        cutoff = now - self._window_sec
        mask = ts >= cutoff

        t = ts[mask]
        self._curve_vout.setData(t, np.array(self._vout)[mask])
        self._curve_vin.setData(t, np.array(self._vin)[mask])
        self._curve_iout.setData(t, np.array(self._iout)[mask])
        self._curve_temp.setData(t, np.array(self._temp)[mask])

        # Remove markers that have scrolled out of the visible window
        still_visible: list[tuple] = []
        for marker in self._markers:
            m_t, _, _, lines, text_item = marker
            if m_t < cutoff:
                for plot, line in zip(self._plots, lines):
                    plot.removeItem(line)
                self._p_volt.removeItem(text_item)
            else:
                still_visible.append(marker)
        self._markers = still_visible

        # Reposition marker labels near top of voltage plot
        if self._markers:
            vr = self._p_volt.viewRange()
            y_top = vr[1][0] + 0.92 * (vr[1][1] - vr[1][0])
            for m_t, _, _, _, text_item in self._markers:
                text_item.setPos(m_t, y_top)

    def _on_pause_toggled(self, checked: bool) -> None:
        self._paused = checked
        self._pause_btn.setText("Resume" if checked else "Pause")

    def _on_window_changed(self, text: str) -> None:
        self._window_sec = WINDOW_OPTIONS.get(text, 600)
        self._data_dirty = True

    def _clear_data(self) -> None:
        self._ts.clear()
        self._vout.clear()
        self._vin.clear()
        self._iout.clear()
        self._temp.clear()
        self._status.clear()
        self._t0 = time.monotonic()
        self._curve_vout.setData([], [])
        self._curve_vin.setData([], [])
        self._curve_iout.setData([], [])
        self._curve_temp.setData([], [])

        # Remove marker graphics
        for _, _, _, lines, text_item in self._markers:
            for plot, line in zip(self._plots, lines):
                plot.removeItem(line)
            self._p_volt.removeItem(text_item)
        self._markers.clear()
        self._marker_data.clear()

    def _export_csv(self) -> None:
        path, _ = QFileDialog.getSaveFileName(
            self,
            "Export CSV",
            f"obc_data_{time.strftime('%Y%m%d_%H%M%S')}.csv",
            "CSV files (*.csv);;All files (*)",
        )
        if not path:
            return
        with open(path, "w", newline="", encoding="utf-8") as f:
            writer = csv.writer(f)
            writer.writerow(
                [
                    "timestamp_s",
                    "Vout_V",
                    "Iout_A",
                    "Vin_V",
                    "Temp_C",
                    "status_flags",
                ]
            )
            for i in range(len(self._ts)):
                writer.writerow([
                    f"{self._ts[i]:.3f}",
                    f"{self._vout[i]:.1f}",
                    f"{self._iout[i]:.1f}",
                    f"{self._vin[i]:.1f}",
                    f"{self._temp[i]:.1f}",
                    f"0x{self._status[i]:02X}",
                ])

            # Append events section
            if self._marker_data:
                writer.writerow([])
                writer.writerow(["# EVENTS"])
                writer.writerow(["timestamp_s", "event_label", "severity"])
                for t, label, sev in self._marker_data:
                    writer.writerow([f"{t:.3f}", label, sev])

File: cpp_app/ui/graph_panel.cpp
#include "ui/graph_panel.h"
#include "ui/theme.h"

#include <QDateTime>
#include <QFileDialog>
#include <QHBoxLayout>
#include <QVBoxLayout>
#include <QFile>
#include <QTextStream>
#include <chrono>
#include <cmath>
#include <algorithm>

static const QMap<QString, int> WINDOW_OPTIONS = {
    {"1 min",  60},
    {"5 min",  300},
    {"10 min", 600},
    {"30 min", 1800},
};

static double monoNowGraph() {
    return std::chrono::duration<double>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}

GraphPanel::GraphPanel(QWidget* parent)
    : QGroupBox("Live Graph", parent)
{
    m_t0 = monoNowGraph();

    auto* layout = new QVBoxLayout(this);

    // Controls row
    auto* ctrlRow = new QHBoxLayout;
    m_pauseBtn = new QPushButton("Pause");
    m_pauseBtn->setCheckable(true);
    connect(m_pauseBtn, &QPushButton::toggled, this, &GraphPanel::onPauseToggled);
    ctrlRow->addWidget(m_pauseBtn);

    m_clearBtn = new QPushButton("Clear");
    connect(m_clearBtn, &QPushButton::clicked, this, &GraphPanel::clearData);
    ctrlRow->addWidget(m_clearBtn);

    ctrlRow->addWidget(new QLabel("Window:"));
    m_windowCombo = new QComboBox;
    m_windowCombo->addItems({"1 min", "5 min", "10 min", "30 min"});
    m_windowCombo->setCurrentText("10 min");
    connect(m_windowCombo, &QComboBox::currentTextChanged, this, &GraphPanel::onWindowChanged);
    ctrlRow->addWidget(m_windowCombo);

    m_exportBtn = new QPushButton("Export CSV");
    connect(m_exportBtn, &QPushButton::clicked, this, &GraphPanel::exportCsv);
    ctrlRow->addWidget(m_exportBtn);
    ctrlRow->addStretch();
    layout->addLayout(ctrlRow);

    // Voltage chart
    m_voltChart = new QChart;
    setupChart(m_voltChart, "Voltage (V)");
    m_voutSeries = new QLineSeries;
    m_voutSeries->setName("Vout");
    m_voutSeries->setColor(QColor(Theme::CYAN));
    m_vinSeries = new QLineSeries;
    m_vinSeries->setName("Vin");
    m_vinSeries->setColor(QColor(Theme::MAGENTA));
    m_voltChart->addSeries(m_voutSeries);
    m_voltChart->addSeries(m_vinSeries);

    m_voltXAxis = new QValueAxis;
    m_voltXAxis->setTitleText("s");
    m_voltXAxis->setLabelsColor(QColor(Theme::TEXT_DIM));
    m_voltXAxis->setGridLineColor(QColor(Theme::BORDER));
    m_voltYAxis = new QValueAxis;
    m_voltYAxis->setTitleText("V");
    m_voltYAxis->setLabelsColor(QColor(Theme::TEXT_DIM));
    m_voltYAxis->setGridLineColor(QColor(Theme::BORDER));
    m_voltChart->addAxis(m_voltXAxis, Qt::AlignBottom);
    m_voltChart->addAxis(m_voltYAxis, Qt::AlignLeft);
    m_voutSeries->attachAxis(m_voltXAxis);
    m_voutSeries->attachAxis(m_voltYAxis);
    m_vinSeries->attachAxis(m_voltXAxis);
    m_vinSeries->attachAxis(m_voltYAxis);

    m_voltView = new QChartView(m_voltChart);
    m_voltView->setRenderHint(QPainter::Antialiasing);
    layout->addWidget(m_voltView, 1);

    // Current chart
    m_currChart = new QChart;
    setupChart(m_currChart, "Current (A)");
    m_ioutSeries = new QLineSeries;
    m_ioutSeries->setName("Iout");
    m_ioutSeries->setColor(QColor(Theme::GREEN));
    m_currChart->addSeries(m_ioutSeries);

    m_currXAxis = new QValueAxis;
    m_currXAxis->setTitleText("s");
    m_currXAxis->setLabelsColor(QColor(Theme::TEXT_DIM));
    m_currXAxis->setGridLineColor(QColor(Theme::BORDER));
    m_currYAxis = new QValueAxis;
    m_currYAxis->setTitleText("A");
    m_currYAxis->setLabelsColor(QColor(Theme::TEXT_DIM));
    m_currYAxis->setGridLineColor(QColor(Theme::BORDER));
    m_currChart->addAxis(m_currXAxis, Qt::AlignBottom);
    m_currChart->addAxis(m_currYAxis, Qt::AlignLeft);
    m_ioutSeries->attachAxis(m_currXAxis);
    m_ioutSeries->attachAxis(m_currYAxis);

    m_currView = new QChartView(m_currChart);
    m_currView->setRenderHint(QPainter::Antialiasing);
    layout->addWidget(m_currView, 1);

    // Temperature chart
    m_tempChart = new QChart;
    setupChart(m_tempChart, QString::fromUtf8("Temperature (\u00b0C)"));
    m_tempSeries = new QLineSeries;
    m_tempSeries->setName("Temp");
    m_tempSeries->setColor(QColor(Theme::ORANGE));
    m_tempChart->addSeries(m_tempSeries);

    m_tempXAxis = new QValueAxis;
    m_tempXAxis->setTitleText("s");
    m_tempXAxis->setLabelsColor(QColor(Theme::TEXT_DIM));
    m_tempXAxis->setGridLineColor(QColor(Theme::BORDER));
    m_tempYAxis = new QValueAxis;
    m_tempYAxis->setTitleText(QString::fromUtf8("\u00b0C"));
    m_tempYAxis->setLabelsColor(QColor(Theme::TEXT_DIM));
    m_tempYAxis->setGridLineColor(QColor(Theme::BORDER));
    m_tempChart->addAxis(m_tempXAxis, Qt::AlignBottom);
    m_tempChart->addAxis(m_tempYAxis, Qt::AlignLeft);
    m_tempSeries->attachAxis(m_tempXAxis);
    m_tempSeries->attachAxis(m_tempYAxis);

    m_tempView = new QChartView(m_tempChart);
    m_tempView->setRenderHint(QPainter::Antialiasing);
    layout->addWidget(m_tempView, 1);

    // Redraw timer (~15 Hz)
    m_redrawTimer = new QTimer(this);
    connect(m_redrawTimer, &QTimer::timeout, this, &GraphPanel::redraw);
    m_redrawTimer->start(66);
}

void GraphPanel::setupChart(QChart* chart, const QString& title) {
    chart->setTitle(title);
    chart->setTitleBrush(QBrush(QColor(Theme::TEXT_DIM)));
    chart->setBackgroundBrush(QBrush(QColor(Theme::BG_DEEP)));
    chart->setPlotAreaBackgroundBrush(QBrush(QColor("#0d1117")));
    chart->setPlotAreaBackgroundVisible(true);
    chart->legend()->setLabelColor(QColor(Theme::TEXT_DIM));
    chart->legend()->setAlignment(Qt::AlignTop);
    chart->setMargins(QMargins(4, 4, 4, 4));
}

void GraphPanel::addPoint(const Message2& msg) {
    double t = monoNowGraph() - m_t0;
    m_ts.push_back(t);
    m_vout.push_back(msg.output_voltage);
    m_vin.push_back(msg.input_voltage);
    m_iout.push_back(msg.output_current);
    m_temp.push_back(msg.temperature);
    m_status.push_back(msg.status.toByte());

    // Enforce max size
    while (m_ts.size() > MAX_POINTS) {
        m_ts.pop_front();
        m_vout.pop_front();
        m_vin.pop_front();
        m_iout.pop_front();
        m_temp.pop_front();
        m_status.pop_front();
    }
    m_dataDirty = true;
}

void GraphPanel::addEventMarker(const QString& label, const QString& severity) {
    double t = monoNowGraph() - m_t0;
    m_markerData.emplace_back(t, label, severity);
}

void GraphPanel::redraw() {
    if (m_paused || m_ts.empty() || !m_dataDirty)
        return;
    m_dataDirty = false;

    double now = m_ts.back();
    double cutoff = now - m_windowSec;

    // Find start index
    size_t startIdx = 0;
    for (size_t i = 0; i < m_ts.size(); ++i) {
        if (m_ts[i] >= cutoff) { startIdx = i; break; }
    }

    // Build point lists
    QList<QPointF> voutPts, vinPts, ioutPts, tempPts;
    double vMin = 1e9, vMax = -1e9;
    double iMin = 1e9, iMax = -1e9;
    double tMin = 1e9, tMax = -1e9;

    for (size_t i = startIdx; i < m_ts.size(); ++i) {
        double t = m_ts[i];
        voutPts.append(QPointF(t, m_vout[i]));
        vinPts.append(QPointF(t, m_vin[i]));
        ioutPts.append(QPointF(t, m_iout[i]));
        tempPts.append(QPointF(t, m_temp[i]));

        double vmax_i = std::max(m_vout[i], m_vin[i]);
        double vmin_i = std::min(m_vout[i], m_vin[i]);
        vMin = std::min(vMin, vmin_i);
        vMax = std::max(vMax, vmax_i);
        iMin = std::min(iMin, m_iout[i]);
        iMax = std::max(iMax, m_iout[i]);
        tMin = std::min(tMin, m_temp[i]);
        tMax = std::max(tMax, m_temp[i]);
    }

    m_voutSeries->replace(voutPts);
    m_vinSeries->replace(vinPts);
    m_ioutSeries->replace(ioutPts);
    m_tempSeries->replace(tempPts);

    // Update axes
    double margin = 5.0;
    m_voltXAxis->setRange(cutoff, now);
    m_voltYAxis->setRange(vMin - margin, vMax + margin);
    m_currXAxis->setRange(cutoff, now);
    m_currYAxis->setRange(iMin - margin, iMax + margin);
    m_tempXAxis->setRange(cutoff, now);
    m_tempYAxis->setRange(tMin - margin, tMax + margin);
}

void GraphPanel::onPauseToggled(bool checked) {
    m_paused = checked;
    m_pauseBtn->setText(checked ? "Resume" : "Pause");
}

void GraphPanel::onWindowChanged(const QString& text) {
    auto it = WINDOW_OPTIONS.find(text);
    m_windowSec = (it != WINDOW_OPTIONS.end()) ? it.value() : 600;
    m_dataDirty = true;
}

void GraphPanel::clearData() {
    m_ts.clear();
    m_vout.clear();
    m_vin.clear();
    m_iout.clear();
    m_temp.clear();
    m_status.clear();
    m_markerData.clear();
    m_t0 = monoNowGraph();

    m_voutSeries->clear();
    m_vinSeries->clear();
    m_ioutSeries->clear();
    m_tempSeries->clear();
}

void GraphPanel::exportCsv() {
    QString path = QFileDialog::getSaveFileName(
        this, "Export CSV",
        QString("obc_data_%1.csv").arg(QDateTime::currentDateTime().toString("yyyyMMdd_HHmmss")),
        "CSV files (*.csv);;All files (*)");
    if (path.isEmpty()) return;

    QFile file(path);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text))
        return;

    QTextStream out(&file);
    out << "timestamp_s,Vout_V,Iout_A,Vin_V,Temp_C,status_flags\n";
    for (size_t i = 0; i < m_ts.size(); ++i) {
        out << QString::number(m_ts[i], 'f', 3) << ","
            << QString::number(m_vout[i], 'f', 1) << ","
            << QString::number(m_iout[i], 'f', 1) << ","
            << QString::number(m_vin[i], 'f', 1) << ","
            << QString::number(m_temp[i], 'f', 1) << ","
            << QString("0x%1").arg(m_status[i], 2, 16, QChar('0')) << "\n";
    }

    if (!m_markerData.empty()) {
        out << "\n# EVENTS\n";
        out << "timestamp_s,event_label,severity\n";
        for (const auto& [t, label, sev] : m_markerData) {
            out << QString::number(t, 'f', 3) << "," << label << "," << sev << "\n";
        }
    }
    file.close();
}
